Pincode based distance and Enhancements for Streamlit App

manujosephv opened the pull request:
key changes:
1. Added pin code based distance as another column
2. made some changes in the streamlit app - some performance and some the way it is displayed
3. filtered out centers with all slots as zero
4. made email trigger to mail that there is no slot, instead of not sending any email at all.

manujosephv committed:
Update availability.py

manujosephv committed:
Merge pull request #1 from manujosephv/patch-1

Update availability.py

manujosephv committed:
Update cron.yaml

manujosephv committed:
-- improved streamlit app
-- added pincode vased distance calculation
-- added empty email if no slots are available

manujosephv committed:
-- changed next_n_days to 1

manujosephv committed:
-- added download button

aahnik commented:
![image](https://user-images.githubusercontent.com/66209958/117464773-9d954180-af6e-11eb-9c4a-088375d07b18.png)

love this @manujosephv 

this fixes #11, as age is logical.

aahnik commented:
but getting 

<img width="500" src="https://user-images.githubusercontent.com/66209958/117464891-bbfb3d00-af6e-11eb-8d01-e5e0300b54cc.png" alt="error-screenshot">


<details>
<summary> 👉 full traceback </summary>


```shell
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/streamlit/script_runner.py", line 337, in _run_script
    exec(code, module.__dict__)
File "/home/aahnik/Desktop/vaccine_availability/streamlit_app.py", line 77, in <module>
    main()
File "/home/aahnik/Desktop/vaccine_availability/streamlit_app.py", line 53, in main
    df = cached_availability(next_n_days, district_ids, min_age_limit, pincode_search, free_paid)
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/streamlit/caching.py", line 573, in wrapped_func
    return get_or_create_cached_value()
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/streamlit/caching.py", line 557, in get_or_create_cached_value
    return_value = func(*args, **kwargs)
File "/home/aahnik/Desktop/vaccine_availability/streamlit_app.py", line 10, in cached_availability
    df= get_availability(next_n_days, district_ids, min_age_limit, pincode_search)
File "/home/aahnik/Desktop/vaccine_availability/availability.py", line 51, in get_availability
    data = get_data(URL)
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/cachetools/func.py", line 66, in wrapper
    v = func(*args, **kwargs)
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/retry/api.py", line 73, in retry_decorator
    return __retry_internal(partial(f, *args, **kwargs), exceptions, tries, delay, max_delay, backoff, jitter,
File "/home/aahnik/Desktop/vaccine_availability/.venv/lib/python3.9/site-packages/retry/api.py", line 33, in __retry_internal
    return f()
File "/home/aahnik/Desktop/vaccine_availability/availability.py", line 37, in get_data
    data = json.loads(response.text)['centers']
File "/usr/lib/python3.9/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
File "/usr/lib/python3.9/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
File "/usr/lib/python3.9/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
```

</details>

manujosephv committed:
-- fixed api pull error
-- added options to show empty slots
-- added option to send/not send empty emails

manujosephv committed:
-- added upstream changes for merge conflicts

manujosephv committed:
Merge branch 'main' into upstream_merge

manujosephv commented:
@bhavsarpratik have resolved the merge conflicts. And fixed the api pull issue. It is now working

manujosephv committed:
-- removed redundant headers from my update

manujosephv commented:
on more investigation... the requests are going through from my local, but not from the Github Action. Not sure why

aahnik commented:
hi @manujosephv, the JSON decode error is resolved after the changes. thanks!

![image](https://user-images.githubusercontent.com/66209958/117524102-b850d000-afd9-11eb-8613-b8c7c555c0d2.png)

46 is greater than 45, so it should show all slots with min age 45 ... right ?

it shows nothing,

![image](https://user-images.githubusercontent.com/66209958/117524122-cf8fbd80-afd9-11eb-8ab0-9a2a43b08da1.png)

but if I do this,

![image](https://user-images.githubusercontent.com/66209958/117524132-dae2e900-afd9-11eb-95f9-f2d1f13b011b.png)


then slots are shown.

![image](https://user-images.githubusercontent.com/66209958/117524141-e3d3ba80-afd9-11eb-9ccb-16f82bc125de.png)

aahnik commented:
without the  "show full slots" being checked, nothing is being shown ever

![image](https://user-images.githubusercontent.com/66209958/117524206-2a291980-afda-11eb-9c59-1eb3618cc49e.png)

aahnik commented:
hi @manujosephv, it would be more logical to have a download pdf button. 


```python
df.to_html() # will give HTML of the dataframe

import pdfkit as pdf

# pip install pdfkit
# sudo apt install wkhtmltopdf

pdf.from_file("/path/to/html.html","/path/to/pdf.pdf")
```

the user of this app will be common human, and csv is for machines to read. if a programmer wants data, they will take from the official API, not after downloading csv from this app.
